In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv("../data/processed/cars_clean.csv")

In [3]:
df.head()

,model,year,price,transmission,mileage,fuelType,tax,mpg,engineSize,Make,age,price_log
0,T-Roc,2019,25000,Automatic,13904,Diesel,145,49.6,2.0,VW,7,10.126671
1,T-Roc,2019,26883,Automatic,4562,Diesel,145,49.6,2.0,VW,7,10.199287
2,T-Roc,2019,20000,Manual,7414,Diesel,145,50.4,2.0,VW,7,9.903538
3,T-Roc,2019,33492,Automatic,4825,Petrol,145,32.5,2.0,VW,7,10.419092
4,T-Roc,2019,22900,Semi-Auto,6500,Petrol,150,39.8,1.5,VW,7,10.038936


In [4]:
df.shape

(97675, 12)

In [5]:
df.dtypes

model               str
year              int64
price             int64
transmission        str
mileage           int64
fuelType            str
tax               int64
mpg             float64
engineSize      float64
Make                str
age               int64
price_log       float64
dtype: object

In [6]:
df["mileage_per_year"] = df["mileage"] / df["age"].replace(0, 1)

df[["mileage", "age", "mileage_per_year"]].head()

,mileage,age,mileage_per_year
0,13904,7,1986.285714
1,4562,7,651.714286
2,7414,7,1059.142857
3,4825,7,689.285714
4,6500,7,928.571429


In [7]:
df["mileage_per_year"].describe()

count    97675.000000
mean      2355.866634
std       1782.271570
min          0.040000
25%       1000.000000
50%       2020.000000
75%       3322.111111
max      24049.400000
Name: mileage_per_year, dtype: float64

In [8]:
df["brand_model"] = df["Make"] + "_" + df["model"]

df[["Make", "model", "brand_model"]].head()

,Make,model,brand_model
0,VW,T-Roc,VW_T-Roc
1,VW,T-Roc,VW_T-Roc
2,VW,T-Roc,VW_T-Roc
3,VW,T-Roc,VW_T-Roc
4,VW,T-Roc,VW_T-Roc


In [9]:
df[["age", "mileage_per_year", "brand_model"]].head()

,age,mileage_per_year,brand_model
0,7,1986.285714,VW_T-Roc
1,7,651.714286,VW_T-Roc
2,7,1059.142857,VW_T-Roc
3,7,689.285714,VW_T-Roc
4,7,928.571429,VW_T-Roc


In [10]:
df[[
    "year",
    "mileage",
    "tax",
    "mpg",
    "engineSize",
    "age",
    "mileage_per_year",
    "price"
]].corr()["price"].sort_values(ascending=False)

price               1.000000
engineSize          0.647807
year                0.494959
tax                 0.307016
mpg                -0.294322
mileage_per_year   -0.391917
mileage            -0.417887
age                -0.494959
Name: price, dtype: float64

In [11]:
df[["Make", "model", "transmission", "fuelType"]].nunique()

Make              9
model           195
transmission      4
fuelType          5
dtype: int64

In [12]:
df["model"].value_counts().tail(10)

model
230                1
220                1
200                1
Accent             1
Amica              1
Ranger             1
Escort             1
Transit Tourneo    1
A2                 1
RS7                1
Name: count, dtype: int64

In [13]:
df.to_csv("../data/processed/cars_features.csv", index=False)

In [14]:
df.shape

(97675, 14)

In [15]:
df["Make"].unique()

<StringArray>
['VW', 'vauxhall', 'toyota', 'skoda', 'merc', 'hyundi', 'ford', 'BMW', 'Audi']
Length: 9, dtype: str

In [16]:
df["Make"] = df["Make"].str.strip().str.lower()

In [17]:
df["Make"].unique()

<StringArray>
['vw', 'vauxhall', 'toyota', 'skoda', 'merc', 'hyundi', 'ford', 'bmw', 'audi']
Length: 9, dtype: str

In [18]:
brand_map = {
    "vw": "volkswagen",
    "vauxhall": "vauxhall",
    "toyota": "toyota",
    "skoda": "skoda",
    "merc": "mercedes-benz",
    "hyundi": "hyundai",
    "ford": "ford",
    "bmw": "bmw",
    "audi": "audi"
}

df["Make"] = df["Make"].replace(brand_map)

In [19]:
df["Make"] = df["Make"].str.strip()

In [20]:
df["Make"].unique()

<StringArray>
[   'volkswagen',      'vauxhall',        'toyota',         'skoda',
 'mercedes-benz',       'hyundai',          'ford',           'bmw',
          'audi']
Length: 9, dtype: str

In [21]:
print(df["Make"].value_counts())

Make
ford             17810
volkswagen       14892
vauxhall         13257
mercedes-benz    12855
bmw              10650
audi             10565
toyota            6688
skoda             6188
hyundai           4770
Name: count, dtype: int64


In [22]:
df["brand_model"] = df["Make"] + "_" + df["model"].str.lower()

In [23]:
df[["Make", "model", "brand_model"]].head(10)

,Make,model,brand_model
0,volkswagen,T-Roc,volkswagen_t-roc
1,volkswagen,T-Roc,volkswagen_t-roc
2,volkswagen,T-Roc,volkswagen_t-roc
3,volkswagen,T-Roc,volkswagen_t-roc
4,volkswagen,T-Roc,volkswagen_t-roc
5,volkswagen,T-Roc,volkswagen_t-roc
6,volkswagen,T-Roc,volkswagen_t-roc
7,volkswagen,T-Roc,volkswagen_t-roc
8,volkswagen,T-Roc,volkswagen_t-roc
9,volkswagen,T-Roc,volkswagen_t-roc


In [24]:
df.shape

(97675, 14)

In [25]:
df.isna().sum()

model               0
year                0
price               0
transmission        0
mileage             0
fuelType            0
tax                 0
mpg                 0
engineSize          0
Make                0
age                 0
price_log           0
mileage_per_year    0
brand_model         0
dtype: int64

In [26]:
df.to_csv("../data/processed/cars_features.csv", index=False)

In [27]:
df.shape

(97675, 14)

## Feature Engineering Summary

The cleaned dataset was enhanced with two meaningful features:

* **`mileage_per_year`** — captures vehicle usage relative to its age.
* **`brand_model`** — combines the standardized brand and model into a specific vehicle identifier.

The final feature dataset contains **97,675 records and 14 columns**.

Target-derived fields such as `price_log` and `price_segment` will not be used as input features for price prediction to avoid target leakage.
